File: patterns/go-tests/test-code.go
package calculator

import (
	"errors"
	"math"
)

// Calculator provides basic arithmetic operations.
type Calculator struct {
	precision int
}

// NewCalculator creates a new Calculator with the specified precision.
func NewCalculator(precision int) *Calculator {
	return &Calculator{precision: precision}
}

// Add returns the sum of two numbers.
func (c *Calculator) Add(a, b float64) float64 {
	return a + b
}

// Subtract returns the difference of two numbers.
func (c *Calculator) Subtract(a, b float64) float64 {
	return a - b
}

// Multiply returns the product of two numbers.
func (c *Calculator) Multiply(a, b float64) float64 {
	return a * b
}

// Divide returns the quotient of two numbers.
// It returns an error if the divisor is zero.
func (c *Calculator) Divide(a, b float64) (float64, error) {
	if b == 0 {
		return 0, errors.New("division by zero")
	}
	return a / b, nil
}

// Power returns a raised to the power of b.
func (c *Calculator) Power(a, b float64) float64 {
	return math.Pow(a, b)
}

// SquareRoot returns the square root of a number.
// It returns an error if the number is negative.
func (c *Calculator) SquareRoot(a float64) (float64, error) {
	if a < 0 {
		return 0, errors.New("cannot compute square root of negative number")
	}
	return math.Sqrt(a), nil
}

// Average returns the average of a slice of numbers.
// It returns an error if the slice is empty.
func Average(numbers []float64) (float64, error) {
	if len(numbers) == 0 {
		return 0, errors.New("cannot compute average of empty slice")
	}

	sum := 0.0
	for _, n := range numbers {
		sum += n
	}
	return sum / float64(len(numbers)), nil
}

// Max returns the maximum value from a slice of numbers.
// It returns an error if the slice is empty.
func Max(numbers []float64) (float64, error) {
	if len(numbers) == 0 {
		return 0, errors.New("cannot find max of empty slice")
	}

	max := numbers[0]
	for _, n := range numbers[1:] {
		if n > max {
			max = n
		}
	}
	return max, nil
}

// Min returns the minimum value from a slice of numbers.
// It returns an error if the slice is empty.
func Min(numbers []float64) (float64, error) {
	if len(numbers) == 0 {
		return 0, errors.New("cannot find min of empty slice")
	}

	min := numbers[0]
	for _, n := range numbers[1:] {
		if n < min {
			min = n
		}
	}
	return min, nil
}

// IsEven reports whether n is even.
func IsEven(n int) bool {
	return n%2 == 0
}

// IsPrime reports whether n is a prime number.
func IsPrime(n int) bool {
	if n <= 1 {
		return false
	}
	if n <= 3 {
		return true
	}
	if n%2 == 0 || n%3 == 0 {
		return false
	}
	for i := 5; i*i <= n; i += 6 {
		if n%i == 0 || n%(i+2) == 0 {
			return false
		}
	}
	return true
}

File: patterns/python-tests/test-code.py
"""Calculator module providing mathematical operations.

This module contains a Calculator class and utility functions for
performing various mathematical calculations.
"""

import math
from typing import List, Optional


class Calculator:
    """A calculator with configurable precision for floating-point operations."""

    def __init__(self, precision: int = 2):
        """Initialize calculator with specified decimal precision.

        Args:
            precision: Number of decimal places for rounding results.
        """
        self.precision = precision

    def add(self, a: float, b: float) -> float:
        """Add two numbers.

        Args:
            a: First number.
            b: Second number.

        Returns:
            Sum of a and b, rounded to calculator's precision.
        """
        return round(a + b, self.precision)

    def subtract(self, a: float, b: float) -> float:
        """Subtract b from a.

        Args:
            a: Number to subtract from.
            b: Number to subtract.

        Returns:
            Difference of a and b, rounded to calculator's precision.
        """
        return round(a - b, self.precision)

    def multiply(self, a: float, b: float) -> float:
        """Multiply two numbers.

        Args:
            a: First number.
            b: Second number.

        Returns:
            Product of a and b, rounded to calculator's precision.
        """
        return round(a * b, self.precision)

    def divide(self, a: float, b: float) -> float:
        """Divide a by b.

        Args:
            a: Dividend.
            b: Divisor.

        Returns:
            Quotient of a and b, rounded to calculator's precision.

        Raises:
            ValueError: If b is zero.
        """
        if b == 0:
            raise ValueError("cannot divide by zero")
        return round(a / b, self.precision)

    def power(self, base: float, exponent: float) -> float:
        """Raise base to the power of exponent.

        Args:
            base: The base number.
            exponent: The exponent.

        Returns:
            base raised to exponent, rounded to calculator's precision.
        """
        return round(math.pow(base, exponent), self.precision)

    def square_root(self, n: float) -> float:
        """Calculate the square root of a number.

        Args:
            n: Number to find square root of.

        Returns:
            Square root of n, rounded to calculator's precision.

        Raises:
            ValueError: If n is negative.
        """
        if n < 0:
            raise ValueError("cannot calculate square root of negative number")
        return round(math.sqrt(n), self.precision)


def average(numbers: List[float]) -> float:
    """Calculate the average of a list of numbers.

    Args:
        numbers: List of numbers to average.

    Returns:
        Arithmetic mean of the numbers.

    Raises:
        ValueError: If the list is empty.
    """
    if not numbers:
        raise ValueError("cannot calculate average of empty list")
    return sum(numbers) / len(numbers)


def maximum(numbers: List[float]) -> float:
    """Find the maximum value in a list of numbers.

    Args:
        numbers: List of numbers to search.

    Returns:
        The largest number in the list.

    Raises:
        ValueError: If the list is empty.
    """
    if not numbers:
        raise ValueError("cannot find maximum of empty list")
    return max(numbers)


def minimum(numbers: List[float]) -> float:
    """Find the minimum value in a list of numbers.

    Args:
        numbers: List of numbers to search.

    Returns:
        The smallest number in the list.

    Raises:
        ValueError: If the list is empty.
    """
    if not numbers:
        raise ValueError("cannot find minimum of empty list")
    return min(numbers)


def is_even(n: int) -> bool:
    """Check if a number is even.

    Args:
        n: Integer to check.

    Returns:
        True if n is even, False otherwise.
    """
    return n % 2 == 0


def is_prime(n: int) -> bool:
    """Check if a number is prime.

    Args:
        n: Integer to check.

    Returns:
        True if n is prime, False otherwise.
    """
    if n < 2:
        return False
    if n == 2:
        return True
    if n % 2 == 0:
        return False
    for i in range(3, int(math.sqrt(n)) + 1, 2):
        if n % i == 0:
            return False
    return True


def factorial(n: int) -> int:
    """Calculate the factorial of a non-negative integer.

    Args:
        n: Non-negative integer.

    Returns:
        n! (n factorial).

    Raises:
        ValueError: If n is negative.
    """
    if n < 0:
        raise ValueError("cannot calculate factorial of negative number")
    if n <= 1:
        return 1
    return n * factorial(n - 1)


def fibonacci(n: int) -> int:
    """Calculate the nth Fibonacci number.

    Args:
        n: Position in Fibonacci sequence (0-indexed).

    Returns:
        The nth Fibonacci number.

    Raises:
        ValueError: If n is negative.
    """
    if n < 0:
        raise ValueError("cannot calculate fibonacci of negative index")
    if n <= 1:
        return n
    a, b = 0, 1
    for _ in range(2, n + 1):
        a, b = b, a + b
    return b
